flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/inputformat/InputFormatCacheLoader.java

```diff
@@ -0,0 +1,159 @@
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+/** {@link CacheLoader} that used {@link InputFormat} for loading data into the cache. */
+public class InputFormatCacheLoader extends CacheLoader {
+    private static final long serialVersionUID = 1L;
+    private static final Logger LOG = LoggerFactory.getLogger(InputFormatCacheLoader.class);
+
+    private final InputFormat<RowData, InputSplit> initialInputFormat;
+    private final GenericRowDataKeySelector keySelector;
+    private final RowDataSerializer cacheEntriesSerializer;
+
+    private transient volatile List<InputSplitCacheLoadTask> cacheLoadTasks;
+    private transient Configuration parameters;
+
+    private volatile boolean isStopped;
+
+    public InputFormatCacheLoader(
+            InputFormat<RowData, ?> initialInputFormat,
+            GenericRowDataKeySelector keySelector,
+            RowDataSerializer cacheEntriesSerializer) {
+        // noinspection unchecked
+        this.initialInputFormat = (InputFormat<RowData, InputSplit>) initialInputFormat;
+        this.keySelector = keySelector;
+        this.cacheEntriesSerializer = cacheEntriesSerializer;
+    }
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        super.open(parameters);
+        this.parameters = parameters;
+        this.initialInputFormat.configure(parameters);
+    }
+
+    @Override
+    protected void reloadCache() throws Exception {
+        InputSplit[] inputSplits = createInputSplits();
+        int numSplits = inputSplits.length;
+        // load data into the another copy of cache
+        // notice: it requires twice more memory, but on the other hand we don't need any blocking
+        // cache has default initialCapacity and loadFactor, but overridden concurrencyLevel
+        ConcurrentHashMap<RowData, Collection<RowData>> newCache =
+                new ConcurrentHashMap<>(16, 0.75f, getConcurrencyLevel(numSplits));
+        this.cacheLoadTasks =
+                Arrays.stream(inputSplits)
+                        .map(split -> createCacheLoadTask(split, newCache))
+                        .collect(Collectors.toList());
+
+        // run first task and start numTasks - 1 threads to run remaining tasks
+        ExecutorService cacheLoadTaskService = null;
+        List<Future<?>> futures = null;
+        if (numSplits > 1) {
+            futures = new ArrayList<>();
+            int numThreads = getConcurrencyLevel(numSplits) - 1;
+            cacheLoadTaskService = Executors.newFixedThreadPool(numThreads);
+            for (int i = 1; i < numSplits; i++) {
+                Future<?> future = cacheLoadTaskService.submit(cacheLoadTasks.get(i));
+                futures.add(future);
+            }
+        }
+        cacheLoadTasks.get(0).run();
```

Comment:
Why do we pick the first task to run synchronously instead of putting all tasks into the executor and wait for all futures to complete?

Reply:
By doing this we don't create one more thread. Parent thread any waits for all tasks to complete, so he can execute one of tasks to not idle.